In [1]:
# Importing Required Packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")
pd.set_option('display.max_columns', None)# To see all the columns of a dataframe
#pd.set_option('display.max_rows', None)

In [2]:
# Loading reduced feature training set
X_train = pd.read_csv('X_train_final.csv')
y_train = pd.read_csv('y_train.final.csv')

In [3]:
# Loading reduced feature test set
X_test = pd.read_csv('X_test_final.csv')
y_test = pd.read_csv('y_test.final.csv')

### Standardizing all the columns, including dummy variables, for SELU activated Dense model

In [4]:
# Importing the Standard scaler from Sklearn
from sklearn.preprocessing import StandardScaler

In [5]:
# Instantiating the Standard Scaler object & fit_transforming the reduced feature training set.
st_scaler = StandardScaler() 
arr = st_scaler.fit_transform(X_train)

In [6]:
# Converting the array to the corresponding dataframe
X_train_scaled = pd.DataFrame(arr, columns=X_train.columns, index=X_train.index)

In [7]:
# Transforming the test set with the same standard scaler object as was used to fit Training set.
arr1 = st_scaler.transform(X_test)

In [8]:
# Converting the array to the corresponding dataframe.
X_test_scaled = pd.DataFrame(arr1, columns=X_test.columns, index=X_test.index)

## Model_11: Keras Dense Model with Equal Neurons/layer, SELU Activation & Tuned learning rate.

In [9]:
# Importing Optuna , Tensorflow & Keras
import optuna , tensorflow as tf

In [10]:
# Importing Keras
from tensorflow import keras

In [11]:
# Importing the Sklearn's roc_auc_score module
from sklearn.metrics import roc_auc_score

In [12]:
# Importing train test split from Sklearn to produce validation set
from sklearn.model_selection import train_test_split

In [13]:
# Performing train_validation split with test size = .15 to provide more Training data to Neural Net
X_train_1, X_val, y_train_1, y_val = train_test_split(X_train_scaled, y_train, test_size=0.15, random_state=42,
                                                          stratify=y_train)

In [14]:
# defining class weight dictionary list for the imbalanced class
cl_weight = [{0:1.0,1:11.38697},{0:1.0,1:12.0},{0:1.0,1:13.0},{0:1.0,1:14.0},{0:1.0,1:15.0},{0:1.0,1:16.0},
             {0:1.0,1:17.0},{0:1.0,1:18.0}]

In [15]:
# Defining best roc_auc score
best_roc_auc = 0

In [16]:
# defining the objective function
def objective_wrappper_keras_eq(X_tr, y_tr, X_val, y_val, cl_w):
    '''
    Optimizes Keras ,neural net with equal no. of neurons in each layer , best parameters 
    on the given training set: X_tr, y_tr using validation set: X_val,y_val & class weights: cl_w
    
    '''

    def objective(trial):
        s = trial.suggest_int('step',2,6)
        def exponential_decay_fn(epoch):
            return 0.01 * 0.1**(epoch /s )
              
        # building model & searching for the best no. of neural units per layer.
        no_units = trial.suggest_int('no._units',100,600)# No. of Neurons in each layer
        model = keras.models.Sequential()
        model.add(keras.layers.InputLayer(input_shape=X_tr.shape[1:]))
        #model.add(keras.layers.BatchNormalization())
        for i in range(trial.suggest_int('number',2,6)):
            model.add(keras.layers.Dense(units=no_units,activation='selu',
                                     kernel_initializer='lecun_normal')) # For self normalization
        model.add(keras.layers.Dense(units=1,activation='sigmoid'))
    
        # Searching for the best learning rate.
        #lr = trial.suggest_float("lr", 1e-5, 1e-1, log=True) 
    
        # Compiling the model
        model.compile(loss="binary_crossentropy", optimizer=keras.optimizers.Nadam(beta_1=0.9, beta_2=0.999),
                      metrics=[tf.keras.metrics.AUC(curve="ROC")])
    
        # Defining the Callbacks
        checkpoint_cb = keras.callbacks.ModelCheckpoint('best_model.h5',save_best_only=True) # 1st Callback
        early_stopping_cb = keras.callbacks.EarlyStopping(patience=4) # 2nd Callback, Stop if validation score doen't improve for
        # 4 epochs        
        lr_scheduler_cb = keras.callbacks.LearningRateScheduler(exponential_decay_fn)# 3rd Callback
        #tensorboard_cb = keras.callbacks.TensorBoard(get_run_logdir()) 
        
        #Fitting the model
        model.fit(X_tr, y_tr, epochs=50, validation_data=(X_val,y_val), batch_size=32,
             class_weight=trial.suggest_categorical('best_weight',cl_w),
             callbacks=[checkpoint_cb, early_stopping_cb, lr_scheduler_cb])#,tensorboard_cb])
    
        # Loading the best model 
        model_best = keras.models.load_model('best_model.h5')
    
        # Computing the roc_auc_score for the validation set
        roc_auc_1 = roc_auc_score(y_val,model_best.predict_proba(X_val))
        global best_roc_auc
        # Updating best_roc_auc score
        if roc_auc_1 > best_roc_auc:
            best_roc_auc = roc_auc_1
            model_best.save('Best_model_Selu_eq_Learn.h5')
        
        return roc_auc_1
    
    return objective

In [17]:
# Defining the evaluation function for study's best parameters
def study_best_score_params(X_tr, y_tr, X_v, y_v, obj_func, cl_w, n_trials=100):
    ''' Computes the best hyper parameters of the Neural net using the Training set(X_tr,y_tr) & Validation set
    (X_v,y_v), class weights cl_w & 
    returns Optuna's study's best score & clasifier parameters'''
    
    study = optuna.create_study(direction='maximize')
    study.optimize(obj_func(X_tr, y_tr, X_v, y_v, cl_w),n_trials)
    best_score = study.best_value
    best_params = study.best_params
    return (best_score,best_params)

In [18]:
# Extracting the best model parameters and best study score
best_study_score,best_study_params = study_best_score_params(X_train_1, y_train_1, X_val, y_val, objective_wrappper_keras_eq, cl_weight,
                                                            n_trials=50)

[I 2021-01-05 00:24:14,144] A new study created in memory with name: no-name-0b21dd42-5b01-4387-8f6e-aa5633f7ef4d


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 25s 130us/sample - loss: 4.6109 - auc: 0.5025 - val_loss: 1.5189 - val_auc: 0.5000
Epoch 2/50
196038/196038 [==============================] - 24s 124us/sample - loss: 1.5165 - auc: 0.4988 - val_loss: 1.5090 - val_auc: 0.5000
Epoch 3/50
196038/196038 [==============================] - 24s 123us/sample - loss: 1.4531 - auc: 0.4985 - val_loss: 1.5077 - val_auc: 0.5000
Epoch 4/50
196038/196038 [==============================] - 24s 124us/sample - loss: 1.4273 - auc: 0.5028 - val_loss: 1.5155 - val_auc: 0.5000
Epoch 5/50
196038/196038 [==============================] - 24s 124us/sample - loss: 1.4171 - auc: 0.4971 - val_loss: 1.4245 - val_auc: 0.5000
Epoch 6/50
196038/196038 [==============================] - 24s 124us/sample - loss: 1.4120 - auc: 0.5016 - val_loss: 1.4099 - val_auc: 0.5000
Epoch 7/50
196038/196038 [===============

[I 2021-01-05 00:31:12,389] Trial 0 finished with value: 0.5 and parameters: {'step': 2, 'no._units': 568, 'number': 5, 'best_weight': {0: 1.0, 1: 14.0}}. Best is trial 0 with value: 0.5.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 18s 93us/sample - loss: 4.0905 - auc_1: 0.6125 - val_loss: 2.9738 - val_auc_1: 0.7284
Epoch 2/50
196038/196038 [==============================] - 17s 88us/sample - loss: 22.3165 - auc_1: 0.6574 - val_loss: 2.2696 - val_auc_1: 0.7352
Epoch 3/50
196038/196038 [==============================] - 17s 88us/sample - loss: 1.4057 - auc_1: 0.7111 - val_loss: 3.1440 - val_auc_1: 0.7388
Epoch 4/50
196038/196038 [==============================] - 17s 88us/sample - loss: 1.4802 - auc_1: 0.7175 - val_loss: 2.0030 - val_auc_1: 0.7491
Epoch 5/50
196038/196038 [==============================] - 17s 88us/sample - loss: 1.3450 - auc_1: 0.7356 - val_loss: 2.2615 - val_auc_1: 0.7481
Epoch 6/50
196038/196038 [==============================] - 17s 88us/sample - loss: 1.3003 - auc_1: 0.7504 - val_loss: 1.9329 - val_auc_1: 0.7490
Epoch 7/50
196038/1960

[I 2021-01-05 00:34:24,887] Trial 1 finished with value: 0.7511888420414058 and parameters: {'step': 6, 'no._units': 543, 'number': 2, 'best_weight': {0: 1.0, 1: 16.0}}. Best is trial 1 with value: 0.7511888420414058.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 19s 97us/sample - loss: 1.9404 - auc_2: 0.5138 - val_loss: 1.6894 - val_auc_2: 0.5000
Epoch 2/50
196038/196038 [==============================] - 19s 97us/sample - loss: 1.5152 - auc_2: 0.4984 - val_loss: 1.4200 - val_auc_2: 0.5000
Epoch 3/50
196038/196038 [==============================] - 19s 97us/sample - loss: 1.4422 - auc_2: 0.4986 - val_loss: 1.4672 - val_auc_2: 0.5000
Epoch 4/50
196038/196038 [==============================] - 19s 96us/sample - loss: 1.4173 - auc_2: 0.5002 - val_loss: 1.3646 - val_auc_2: 0.5000
Epoch 5/50
196038/196038 [==============================] - 19s 95us/sample - loss: 1.3996 - auc_2: 0.5009 - val_loss: 1.3894 - val_auc_2: 0.5000
Epoch 6/50
196038/196038 [==============================] - 19s 94us/sample - loss: 1.3888 - auc_2: 0.5035 - val_loss: 1.3687 - val_auc_2: 0.5000
Epoch 7/50
196038/19603

[I 2021-01-05 00:36:56,247] Trial 2 finished with value: 0.4999842777183825 and parameters: {'step': 5, 'no._units': 351, 'number': 3, 'best_weight': {0: 1.0, 1: 13.0}}. Best is trial 1 with value: 0.7511888420414058.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 20s 101us/sample - loss: 2.2310 - auc_3: 0.5089 - val_loss: 2.7664 - val_auc_3: 0.5001
Epoch 2/50
196038/196038 [==============================] - 19s 95us/sample - loss: 1.6642 - auc_3: 0.5006 - val_loss: 2.1546 - val_auc_3: 0.5000
Epoch 3/50
196038/196038 [==============================] - 19s 95us/sample - loss: 1.5671 - auc_3: 0.4975 - val_loss: 2.4462 - val_auc_3: 0.5000
Epoch 4/50
196038/196038 [==============================] - 19s 95us/sample - loss: 1.5379 - auc_3: 0.4955 - val_loss: 1.6036 - val_auc_3: 0.5000
Epoch 5/50
196038/196038 [==============================] - 19s 95us/sample - loss: 1.5193 - auc_3: 0.5010 - val_loss: 1.4632 - val_auc_3: 0.5000
Epoch 6/50
196038/196038 [==============================] - 19s 95us/sample - loss: 1.5035 - auc_3: 0.5035 - val_loss: 1.6105 - val_auc_3: 0.5000
Epoch 7/50
196038/1960

[I 2021-01-05 00:40:42,670] Trial 3 finished with value: 0.4999842777183825 and parameters: {'step': 6, 'no._units': 401, 'number': 3, 'best_weight': {0: 1.0, 1: 15.0}}. Best is trial 1 with value: 0.7511888420414058.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 23s 116us/sample - loss: 110.6749 - auc_4: 0.5032 - val_loss: 1.8627 - val_auc_4: 0.5000
Epoch 2/50
196038/196038 [==============================] - 22s 110us/sample - loss: 1.4770 - auc_4: 0.4944 - val_loss: 1.5228 - val_auc_4: 0.5000
Epoch 3/50
196038/196038 [==============================] - 22s 111us/sample - loss: 1.4746 - auc_4: 0.5047 - val_loss: 1.4270 - val_auc_4: 0.5000
Epoch 4/50
196038/196038 [==============================] - 22s 111us/sample - loss: 1.4724 - auc_4: 0.4998 - val_loss: 1.5609 - val_auc_4: 0.5000
Epoch 5/50
196038/196038 [==============================] - 22s 110us/sample - loss: 1.4556 - auc_4: 0.4981 - val_loss: 1.4329 - val_auc_4: 0.5000
Epoch 6/50
196038/196038 [==============================] - 22s 110us/sample - loss: 1.4458 - auc_4: 0.4973 - val_loss: 1.7670 - val_auc_4: 0.5000
Epoch 7/50
1960

[I 2021-01-05 00:43:16,977] Trial 4 finished with value: 0.5 and parameters: {'step': 6, 'no._units': 280, 'number': 6, 'best_weight': {0: 1.0, 1: 14.0}}. Best is trial 1 with value: 0.7511888420414058.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 17s 86us/sample - loss: 3.4496 - auc_5: 0.6468 - val_loss: 1.7517 - val_auc_5: 0.7044
Epoch 2/50
196038/196038 [==============================] - 16s 82us/sample - loss: 1.4379 - auc_5: 0.6991 - val_loss: 1.3873 - val_auc_5: 0.7295
Epoch 3/50
196038/196038 [==============================] - 16s 82us/sample - loss: 1.4034 - auc_5: 0.7161 - val_loss: 1.8075 - val_auc_5: 0.7429
Epoch 4/50
196038/196038 [==============================] - 16s 81us/sample - loss: 1.3543 - auc_5: 0.7384 - val_loss: 1.4173 - val_auc_5: 0.7461
Epoch 5/50
196038/196038 [==============================] - 16s 81us/sample - loss: 1.3202 - auc_5: 0.7515 - val_loss: 1.3883 - val_auc_5: 0.7490
Epoch 6/50
196038/196038 [==============================] - 16s 81us/sample - loss: 1.2918 - auc_5: 0.7639 - val_loss: 1.3506 - val_auc_5: 0.7497
Epoch 7/50
196038/19603

[I 2021-01-05 00:46:46,999] Trial 5 finished with value: 0.7496072413057584 and parameters: {'step': 5, 'no._units': 184, 'number': 2, 'best_weight': {0: 1.0, 1: 17.0}}. Best is trial 1 with value: 0.7511888420414058.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 18s 93us/sample - loss: 2.6815 - auc_6: 0.6207 - val_loss: 1.6023 - val_auc_6: 0.7323
Epoch 2/50
196038/196038 [==============================] - 17s 87us/sample - loss: 1.5528 - auc_6: 0.6932 - val_loss: 1.4993 - val_auc_6: 0.7335
Epoch 3/50
196038/196038 [==============================] - 17s 87us/sample - loss: 1.7877 - auc_6: 0.7238 - val_loss: 1.4445 - val_auc_6: 0.7501
Epoch 4/50
196038/196038 [==============================] - 17s 87us/sample - loss: 1.3921 - auc_6: 0.7417 - val_loss: 1.6485 - val_auc_6: 0.7541
Epoch 5/50
196038/196038 [==============================] - 17s 86us/sample - loss: 1.3547 - auc_6: 0.7562 - val_loss: 1.3798 - val_auc_6: 0.7536
Epoch 6/50
196038/196038 [==============================] - 17s 86us/sample - loss: 1.3175 - auc_6: 0.7686 - val_loss: 1.4500 - val_auc_6: 0.7493
Epoch 7/50
196038/19603

[I 2021-01-05 00:49:22,151] Trial 6 finished with value: 0.7538299264111656 and parameters: {'step': 4, 'no._units': 418, 'number': 2, 'best_weight': {0: 1.0, 1: 18.0}}. Best is trial 6 with value: 0.7538299264111656.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 21s 109us/sample - loss: 4.3175 - auc_7: 0.5006 - val_loss: 2.3257 - val_auc_7: 0.5000
Epoch 2/50
196038/196038 [==============================] - 20s 104us/sample - loss: 17.6213 - auc_7: 0.5000 - val_loss: 1.6313 - val_auc_7: 0.5000
Epoch 3/50
196038/196038 [==============================] - 20s 104us/sample - loss: 1.5881 - auc_7: 0.4986 - val_loss: 1.7411 - val_auc_7: 0.5000
Epoch 4/50
196038/196038 [==============================] - 20s 104us/sample - loss: 1.5767 - auc_7: 0.5022 - val_loss: 1.5643 - val_auc_7: 0.5000
Epoch 5/50
196038/196038 [==============================] - 20s 103us/sample - loss: 1.5676 - auc_7: 0.5009 - val_loss: 1.5824 - val_auc_7: 0.5000
Epoch 6/50
196038/196038 [==============================] - 20s 103us/sample - loss: 1.5663 - auc_7: 0.4996 - val_loss: 1.6084 - val_auc_7: 0.5001
Epoch 7/50
19603

[I 2021-01-05 00:59:41,807] Trial 7 finished with value: 0.5000471668448525 and parameters: {'step': 4, 'no._units': 220, 'number': 5, 'best_weight': {0: 1.0, 1: 17.0}}. Best is trial 6 with value: 0.7538299264111656.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 22s 114us/sample - loss: 94.1352 - auc_8: 0.5334 - val_loss: 1.6281 - val_auc_8: 0.5663
Epoch 2/50
196038/196038 [==============================] - 21s 108us/sample - loss: 1.6189 - auc_8: 0.5489 - val_loss: 1.5719 - val_auc_8: 0.5837
Epoch 3/50
196038/196038 [==============================] - 21s 107us/sample - loss: 1.5762 - auc_8: 0.6101 - val_loss: 1.4975 - val_auc_8: 0.6876
Epoch 4/50
196038/196038 [==============================] - 21s 109us/sample - loss: 1.4468 - auc_8: 0.7059 - val_loss: 1.3852 - val_auc_8: 0.7359
Epoch 5/50
196038/196038 [==============================] - 21s 108us/sample - loss: 1.3803 - auc_8: 0.7407 - val_loss: 1.3736 - val_auc_8: 0.7428
Epoch 6/50
196038/196038 [==============================] - 21s 108us/sample - loss: 1.3540 - auc_8: 0.7555 - val_loss: 1.3777 - val_auc_8: 0.7442
Epoch 7/50
19603

[I 2021-01-05 01:03:57,766] Trial 8 finished with value: 0.7456813600029164 and parameters: {'step': 3, 'no._units': 437, 'number': 4, 'best_weight': {0: 1.0, 1: 18.0}}. Best is trial 6 with value: 0.7538299264111656.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 17s 87us/sample - loss: 3.2204 - auc_9: 0.6442 - val_loss: 3.3872 - val_auc_9: 0.7015
Epoch 2/50
196038/196038 [==============================] - 16s 83us/sample - loss: 2.3349 - auc_9: 0.6995 - val_loss: 2.0040 - val_auc_9: 0.7264
Epoch 3/50
196038/196038 [==============================] - 16s 82us/sample - loss: 1.4632 - auc_9: 0.7135 - val_loss: 1.5277 - val_auc_9: 0.7356
Epoch 4/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.4424 - auc_9: 0.7280 - val_loss: 1.3871 - val_auc_9: 0.7516
Epoch 5/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.3895 - auc_9: 0.7423 - val_loss: 1.3736 - val_auc_9: 0.7481
Epoch 6/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.3581 - auc_9: 0.7549 - val_loss: 1.4203 - val_auc_9: 0.7534
Epoch 7/50
196038/19603

[I 2021-01-05 01:06:26,255] Trial 9 finished with value: 0.7489693039592251 and parameters: {'step': 6, 'no._units': 243, 'number': 2, 'best_weight': {0: 1.0, 1: 18.0}}. Best is trial 6 with value: 0.7538299264111656.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 21s 106us/sample - loss: 2.1939 - auc_10: 0.5060 - val_loss: 2.2771 - val_auc_10: 0.5000
Epoch 2/50
196038/196038 [==============================] - 20s 100us/sample - loss: 1.4099 - auc_10: 0.4986 - val_loss: 1.3731 - val_auc_10: 0.5000
Epoch 3/50
196038/196038 [==============================] - 20s 100us/sample - loss: 1.3306 - auc_10: 0.5028 - val_loss: 1.4310 - val_auc_10: 0.5000
Epoch 4/50
196038/196038 [==============================] - 19s 98us/sample - loss: 1.3097 - auc_10: 0.5020 - val_loss: 1.4357 - val_auc_10: 0.5000
Epoch 5/50
196038/196038 [==============================] - 19s 98us/sample - loss: 1.2946 - auc_10: 0.4974 - val_loss: 1.2749 - val_auc_10: 0.5000
Epoch 6/50
196038/196038 [==============================] - 19s 98us/sample - loss: 1.2863 - auc_10: 0.4998 - val_loss: 1.2749 - val_auc_10: 0.5000
Epoch 7/

[I 2021-01-05 01:09:23,182] Trial 10 finished with value: 0.4999842777183825 and parameters: {'step': 3, 'no._units': 482, 'number': 3, 'best_weight': {0: 1.0, 1: 11.38697}}. Best is trial 6 with value: 0.7538299264111656.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 19s 95us/sample - loss: 3.9995 - auc_11: 0.6150 - val_loss: 1.6709 - val_auc_11: 0.7334
Epoch 2/50
196038/196038 [==============================] - 18s 91us/sample - loss: 2.9704 - auc_11: 0.6735 - val_loss: 1.6474 - val_auc_11: 0.7444
Epoch 3/50
196038/196038 [==============================] - 18s 91us/sample - loss: 1.4278 - auc_11: 0.7113 - val_loss: 1.5029 - val_auc_11: 0.7464
Epoch 4/50
196038/196038 [==============================] - 18s 91us/sample - loss: 1.3452 - auc_11: 0.7300 - val_loss: 1.3647 - val_auc_11: 0.7486
Epoch 5/50
196038/196038 [==============================] - 18s 91us/sample - loss: 1.3048 - auc_11: 0.7477 - val_loss: 1.3699 - val_auc_11: 0.7522
Epoch 6/50
196038/196038 [==============================] - 18s 90us/sample - loss: 1.2635 - auc_11: 0.7629 - val_loss: 1.3624 - val_auc_11: 0.7515
Epoch 7/50


[I 2021-01-05 01:12:22,990] Trial 11 finished with value: 0.7516476647159779 and parameters: {'step': 4, 'no._units': 582, 'number': 2, 'best_weight': {0: 1.0, 1: 16.0}}. Best is trial 6 with value: 0.7538299264111656.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 17s 89us/sample - loss: 1.2926 - auc_12: 0.6790 - val_loss: 1.3084 - val_auc_12: 0.7339
Epoch 2/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.6505 - auc_12: 0.7077 - val_loss: 2.0119 - val_auc_12: 0.6476
Epoch 3/50
196038/196038 [==============================] - 16s 84us/sample - loss: 1.3443 - auc_12: 0.7022 - val_loss: 1.1758 - val_auc_12: 0.7406
Epoch 4/50
196038/196038 [==============================] - 16s 84us/sample - loss: 1.1190 - auc_12: 0.7512 - val_loss: 1.1753 - val_auc_12: 0.7512
Epoch 5/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.0958 - auc_12: 0.7628 - val_loss: 1.1806 - val_auc_12: 0.7468
Epoch 6/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.0725 - auc_12: 0.7739 - val_loss: 1.1521 - val_auc_12: 0.7502
Epoch 7/50


[I 2021-01-05 01:15:09,045] Trial 12 finished with value: 0.7501665785025511 and parameters: {'step': 4, 'no._units': 116, 'number': 2, 'best_weight': {0: 1.0, 1: 12.0}}. Best is trial 6 with value: 0.7538299264111656.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 20s 104us/sample - loss: 6.6861 - auc_13: 0.5448 - val_loss: 2.7305 - val_auc_13: 0.5000
Epoch 2/50
196038/196038 [==============================] - 19s 98us/sample - loss: 1.7193 - auc_13: 0.4988 - val_loss: 1.6147 - val_auc_13: 0.5003
Epoch 3/50
196038/196038 [==============================] - 19s 98us/sample - loss: 1.6898 - auc_13: 0.5034 - val_loss: 1.6189 - val_auc_13: 0.5002
Epoch 4/50
196038/196038 [==============================] - 19s 98us/sample - loss: 2.7598 - auc_13: 0.5045 - val_loss: 1.5836 - val_auc_13: 0.5000
Epoch 5/50
196038/196038 [==============================] - 19s 98us/sample - loss: 1.5250 - auc_13: 0.5064 - val_loss: 1.5192 - val_auc_13: 0.5000
Epoch 6/50
196038/196038 [==============================] - 19s 98us/sample - loss: 1.5235 - auc_13: 0.5020 - val_loss: 1.5038 - val_auc_13: 0.5000
Epoch 7/50

[I 2021-01-05 01:21:17,690] Trial 13 finished with value: 0.5 and parameters: {'step': 3, 'no._units': 512, 'number': 3, 'best_weight': {0: 1.0, 1: 16.0}}. Best is trial 6 with value: 0.7538299264111656.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 24s 121us/sample - loss: 6.2679 - auc_14: 0.5046 - val_loss: 1.9535 - val_auc_14: 0.5004
Epoch 2/50
196038/196038 [==============================] - 23s 116us/sample - loss: 2.1116 - auc_14: 0.5020 - val_loss: 3.8517 - val_auc_14: 0.5000
Epoch 3/50
196038/196038 [==============================] - 23s 119us/sample - loss: 1.7397 - auc_14: 0.4972 - val_loss: 1.7867 - val_auc_14: 0.5000
Epoch 4/50
196038/196038 [==============================] - 23s 117us/sample - loss: 1.6810 - auc_14: 0.5011 - val_loss: 1.5859 - val_auc_14: 0.5000
Epoch 5/50
196038/196038 [==============================] - 23s 116us/sample - loss: 1.6536 - auc_14: 0.4987 - val_loss: 1.9124 - val_auc_14: 0.5000
Epoch 6/50
196038/196038 [==============================] - 23s 116us/sample - loss: 1.6377 - auc_14: 0.5007 - val_loss: 1.6750 - val_auc_14: 0.5000
Epoch

[I 2021-01-05 01:24:24,023] Trial 14 finished with value: 0.5 and parameters: {'step': 5, 'no._units': 589, 'number': 4, 'best_weight': {0: 1.0, 1: 18.0}}. Best is trial 6 with value: 0.7538299264111656.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 18s 91us/sample - loss: 2.2470 - auc_15: 0.6300 - val_loss: 1.8587 - val_auc_15: 0.7339
Epoch 2/50
196038/196038 [==============================] - 17s 86us/sample - loss: 1.5198 - auc_15: 0.6970 - val_loss: 1.4103 - val_auc_15: 0.7454
Epoch 3/50
196038/196038 [==============================] - 17s 86us/sample - loss: 1.3566 - auc_15: 0.7257 - val_loss: 1.9323 - val_auc_15: 0.7483
Epoch 4/50
196038/196038 [==============================] - 17s 87us/sample - loss: 1.3095 - auc_15: 0.7436 - val_loss: 1.3157 - val_auc_15: 0.7450
Epoch 5/50
196038/196038 [==============================] - 17s 86us/sample - loss: 1.2766 - auc_15: 0.7576 - val_loss: 1.3364 - val_auc_15: 0.7484
Epoch 6/50
196038/196038 [==============================] - 17s 86us/sample - loss: 1.2443 - auc_15: 0.7705 - val_loss: 1.3160 - val_auc_15: 0.7506
Epoch 7/50


[I 2021-01-05 01:26:41,908] Trial 15 finished with value: 0.7453165332676387 and parameters: {'step': 4, 'no._units': 346, 'number': 2, 'best_weight': {0: 1.0, 1: 16.0}}. Best is trial 6 with value: 0.7538299264111656.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 21s 106us/sample - loss: 2.5653 - auc_16: 0.4997 - val_loss: 2.6713 - val_auc_16: 0.5000
Epoch 2/50
196038/196038 [==============================] - 20s 100us/sample - loss: 1.4436 - auc_16: 0.5011 - val_loss: 1.3915 - val_auc_16: 0.5000
Epoch 3/50
196038/196038 [==============================] - 20s 100us/sample - loss: 1.3922 - auc_16: 0.5002 - val_loss: 1.3889 - val_auc_16: 0.5000
Epoch 4/50
196038/196038 [==============================] - 20s 100us/sample - loss: 1.3747 - auc_16: 0.5014 - val_loss: 1.3972 - val_auc_16: 0.5001
Epoch 5/50
196038/196038 [==============================] - 20s 100us/sample - loss: 1.3711 - auc_16: 0.4996 - val_loss: 1.3776 - val_auc_16: 0.5001
Epoch 6/50
196038/196038 [==============================] - 20s 100us/sample - loss: 1.3628 - auc_16: 0.4990 - val_loss: 1.3614 - val_auc_16: 0.5001
Epoch

[I 2021-01-05 01:31:19,854] Trial 16 finished with value: 0.5001415005345575 and parameters: {'step': 2, 'no._units': 458, 'number': 3, 'best_weight': {0: 1.0, 1: 13.0}}. Best is trial 6 with value: 0.7538299264111656.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 18s 90us/sample - loss: 16.5123 - auc_17: 0.6349 - val_loss: 1.5516 - val_auc_17: 0.7162
Epoch 2/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.4795 - auc_17: 0.7083 - val_loss: 1.4607 - val_auc_17: 0.7366
Epoch 3/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.4240 - auc_17: 0.7252 - val_loss: 1.3844 - val_auc_17: 0.7447
Epoch 4/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.3749 - auc_17: 0.7452 - val_loss: 1.3897 - val_auc_17: 0.7505
Epoch 5/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.3379 - auc_17: 0.7609 - val_loss: 1.4204 - val_auc_17: 0.7512
Epoch 6/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.3099 - auc_17: 0.7707 - val_loss: 1.4281 - val_auc_17: 0.7515
Epoch 7/50

[I 2021-01-05 01:33:18,958] Trial 17 finished with value: 0.744697335439725 and parameters: {'step': 3, 'no._units': 385, 'number': 2, 'best_weight': {0: 1.0, 1: 18.0}}. Best is trial 6 with value: 0.7538299264111656.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 24s 122us/sample - loss: 4.0745 - auc_18: 0.5013 - val_loss: 2.5943 - val_auc_18: 0.5000
Epoch 2/50
196038/196038 [==============================] - 23s 115us/sample - loss: 1.5159 - auc_18: 0.4981 - val_loss: 2.0576 - val_auc_18: 0.5000
Epoch 3/50
196038/196038 [==============================] - 23s 115us/sample - loss: 1.3913 - auc_18: 0.5001 - val_loss: 1.2771 - val_auc_18: 0.5000
Epoch 4/50
196038/196038 [==============================] - 23s 115us/sample - loss: 1.3535 - auc_18: 0.4983 - val_loss: 1.3930 - val_auc_18: 0.5000
Epoch 5/50
196038/196038 [==============================] - 22s 115us/sample - loss: 1.3292 - auc_18: 0.5027 - val_loss: 1.3196 - val_auc_18: 0.5000
Epoch 6/50
196038/196038 [==============================] - 23s 115us/sample - loss: 1.3143 - auc_18: 0.4993 - val_loss: 1.2844 - val_auc_18: 0.5000
Epoch

[I 2021-01-05 01:35:59,861] Trial 18 finished with value: 0.5000157222816175 and parameters: {'step': 5, 'no._units': 590, 'number': 4, 'best_weight': {0: 1.0, 1: 11.38697}}. Best is trial 6 with value: 0.7538299264111656.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 25s 127us/sample - loss: 82.2604 - auc_19: 0.4997 - val_loss: 2.2838 - val_auc_19: 0.5000
Epoch 2/50
196038/196038 [==============================] - 23s 120us/sample - loss: 1.6047 - auc_19: 0.5023 - val_loss: 1.7063 - val_auc_19: 0.5000
Epoch 3/50
196038/196038 [==============================] - 23s 120us/sample - loss: 1.5482 - auc_19: 0.4982 - val_loss: 1.4730 - val_auc_19: 0.5000
Epoch 4/50
196038/196038 [==============================] - 24s 120us/sample - loss: 1.5173 - auc_19: 0.5010 - val_loss: 1.5052 - val_auc_19: 0.5000
Epoch 5/50
196038/196038 [==============================] - 24s 120us/sample - loss: 1.4983 - auc_19: 0.5011 - val_loss: 1.8861 - val_auc_19: 0.5000
Epoch 6/50
196038/196038 [==============================] - 24s 121us/sample - loss: 1.4876 - auc_19: 0.4995 - val_loss: 1.4879 - val_auc_19: 0.5000
Epoc

[I 2021-01-05 01:47:03,820] Trial 19 finished with value: 0.5 and parameters: {'step': 4, 'no._units': 521, 'number': 5, 'best_weight': {0: 1.0, 1: 15.0}}. Best is trial 6 with value: 0.7538299264111656.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 18s 94us/sample - loss: 1.7434 - auc_20: 0.5290 - val_loss: 1.6259 - val_auc_20: 0.5000
Epoch 2/50
196038/196038 [==============================] - 18s 90us/sample - loss: 1.4445 - auc_20: 0.4994 - val_loss: 1.3795 - val_auc_20: 0.5000
Epoch 3/50
196038/196038 [==============================] - 18s 89us/sample - loss: 1.3634 - auc_20: 0.4973 - val_loss: 1.3494 - val_auc_20: 0.5000
Epoch 4/50
196038/196038 [==============================] - 18s 90us/sample - loss: 1.3456 - auc_20: 0.4984 - val_loss: 1.3125 - val_auc_20: 0.5000
Epoch 5/50
196038/196038 [==============================] - 18s 90us/sample - loss: 1.3317 - auc_20: 0.5008 - val_loss: 1.3179 - val_auc_20: 0.5000
Epoch 6/50
196038/196038 [==============================] - 17s 89us/sample - loss: 1.3253 - auc_20: 0.5024 - val_loss: 1.3222 - val_auc_20: 0.5000
Epoch 7/50


[I 2021-01-05 01:51:11,985] Trial 20 finished with value: 0.5 and parameters: {'step': 4, 'no._units': 292, 'number': 3, 'best_weight': {0: 1.0, 1: 12.0}}. Best is trial 6 with value: 0.7538299264111656.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 18s 90us/sample - loss: 6.4481 - auc_21: 0.6110 - val_loss: 2.9836 - val_auc_21: 0.7144
Epoch 2/50
196038/196038 [==============================] - 17s 86us/sample - loss: 1.6002 - auc_21: 0.6658 - val_loss: 1.3839 - val_auc_21: 0.7383
Epoch 3/50
196038/196038 [==============================] - 17s 86us/sample - loss: 1.4466 - auc_21: 0.7033 - val_loss: 1.8699 - val_auc_21: 0.7485
Epoch 4/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.3698 - auc_21: 0.7250 - val_loss: 1.7403 - val_auc_21: 0.7480
Epoch 5/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.3215 - auc_21: 0.7429 - val_loss: 1.3082 - val_auc_21: 0.7422
Epoch 6/50
196038/196038 [==============================] - 17s 86us/sample - loss: 1.2820 - auc_21: 0.7579 - val_loss: 1.3159 - val_auc_21: 0.7493
Epoch 7/50


[I 2021-01-05 01:53:45,295] Trial 21 finished with value: 0.7425043670565185 and parameters: {'step': 5, 'no._units': 541, 'number': 2, 'best_weight': {0: 1.0, 1: 16.0}}. Best is trial 6 with value: 0.7538299264111656.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 18s 93us/sample - loss: 4.7534 - auc_22: 0.6107 - val_loss: 2.4777 - val_auc_22: 0.7206
Epoch 2/50
196038/196038 [==============================] - 17s 88us/sample - loss: 1.8623 - auc_22: 0.6890 - val_loss: 1.3173 - val_auc_22: 0.7379
Epoch 3/50
196038/196038 [==============================] - 17s 88us/sample - loss: 1.3593 - auc_22: 0.7238 - val_loss: 1.3400 - val_auc_22: 0.7471
Epoch 4/50
196038/196038 [==============================] - 17s 88us/sample - loss: 1.3030 - auc_22: 0.7464 - val_loss: 1.5077 - val_auc_22: 0.7422
Epoch 5/50
196038/196038 [==============================] - 17s 88us/sample - loss: 1.2670 - auc_22: 0.7613 - val_loss: 1.4230 - val_auc_22: 0.7516
Epoch 6/50
196038/196038 [==============================] - 17s 88us/sample - loss: 1.2392 - auc_22: 0.7727 - val_loss: 1.3057 - val_auc_22: 0.7499
Epoch 7/50


[I 2021-01-05 01:57:14,324] Trial 22 finished with value: 0.7504404152771896 and parameters: {'step': 3, 'no._units': 501, 'number': 2, 'best_weight': {0: 1.0, 1: 16.0}}. Best is trial 6 with value: 0.7538299264111656.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 18s 93us/sample - loss: 6.7135 - auc_23: 0.5755 - val_loss: 3.9780 - val_auc_23: 0.6232
Epoch 2/50
196038/196038 [==============================] - 17s 88us/sample - loss: 1.7404 - auc_23: 0.6557 - val_loss: 1.5784 - val_auc_23: 0.7296
Epoch 3/50
196038/196038 [==============================] - 17s 88us/sample - loss: 1.4676 - auc_23: 0.6953 - val_loss: 1.6627 - val_auc_23: 0.7368
Epoch 4/50
196038/196038 [==============================] - 17s 88us/sample - loss: 1.3897 - auc_23: 0.7171 - val_loss: 2.1479 - val_auc_23: 0.7481
Epoch 5/50
196038/196038 [==============================] - 17s 88us/sample - loss: 1.3382 - auc_23: 0.7357 - val_loss: 1.4144 - val_auc_23: 0.7480
Epoch 6/50
196038/196038 [==============================] - 17s 88us/sample - loss: 1.2954 - auc_23: 0.7522 - val_loss: 1.3051 - val_auc_23: 0.7484
Epoch 7/50


[I 2021-01-05 02:00:44,177] Trial 23 finished with value: 0.7482348600620113 and parameters: {'step': 6, 'no._units': 553, 'number': 2, 'best_weight': {0: 1.0, 1: 16.0}}. Best is trial 6 with value: 0.7538299264111656.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 21s 107us/sample - loss: 3.2651 - auc_24: 0.5029 - val_loss: 2.9235 - val_auc_24: 0.5000
Epoch 2/50
196038/196038 [==============================] - 20s 101us/sample - loss: 1.8159 - auc_24: 0.5004 - val_loss: 1.8229 - val_auc_24: 0.5000
Epoch 3/50
196038/196038 [==============================] - 20s 101us/sample - loss: 1.6438 - auc_24: 0.5015 - val_loss: 1.8990 - val_auc_24: 0.5000
Epoch 4/50
196038/196038 [==============================] - 20s 100us/sample - loss: 1.5970 - auc_24: 0.4964 - val_loss: 1.9326 - val_auc_24: 0.5000
Epoch 5/50
196038/196038 [==============================] - 20s 101us/sample - loss: 1.5658 - auc_24: 0.4999 - val_loss: 1.6887 - val_auc_24: 0.5000
Epoch 6/50
196038/196038 [==============================] - 20s 101us/sample - loss: 1.5478 - auc_24: 0.5013 - val_loss: 1.6296 - val_auc_24: 0.5000
Epoch

[I 2021-01-05 02:07:21,274] Trial 24 finished with value: 0.5 and parameters: {'step': 5, 'no._units': 599, 'number': 3, 'best_weight': {0: 1.0, 1: 16.0}}. Best is trial 6 with value: 0.7538299264111656.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 18s 90us/sample - loss: 4.1023 - auc_25: 0.6229 - val_loss: 3.4908 - val_auc_25: 0.7169
Epoch 2/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.5835 - auc_25: 0.6890 - val_loss: 2.1023 - val_auc_25: 0.7294
Epoch 3/50
196038/196038 [==============================] - 17s 86us/sample - loss: 1.3919 - auc_25: 0.7193 - val_loss: 1.5165 - val_auc_25: 0.7414
Epoch 4/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.3267 - auc_25: 0.7360 - val_loss: 1.4270 - val_auc_25: 0.7505
Epoch 5/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.2888 - auc_25: 0.7531 - val_loss: 1.4612 - val_auc_25: 0.7500
Epoch 6/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.2546 - auc_25: 0.7667 - val_loss: 1.4172 - val_auc_25: 0.7487
Epoch 7/50


[I 2021-01-05 02:10:29,364] Trial 25 finished with value: 0.7499886234403333 and parameters: {'step': 4, 'no._units': 432, 'number': 2, 'best_weight': {0: 1.0, 1: 16.0}}. Best is trial 6 with value: 0.7538299264111656.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 18s 93us/sample - loss: 3.6749 - auc_26: 0.5682 - val_loss: 3.0846 - val_auc_26: 0.5844
Epoch 2/50
196038/196038 [==============================] - 17s 86us/sample - loss: 1.5338 - auc_26: 0.6679 - val_loss: 1.5726 - val_auc_26: 0.7414
Epoch 3/50
196038/196038 [==============================] - 17s 86us/sample - loss: 1.3841 - auc_26: 0.7389 - val_loss: 1.4483 - val_auc_26: 0.7512
Epoch 4/50
196038/196038 [==============================] - 17s 87us/sample - loss: 1.3333 - auc_26: 0.7600 - val_loss: 1.3690 - val_auc_26: 0.7530
Epoch 5/50
196038/196038 [==============================] - 17s 87us/sample - loss: 1.3113 - auc_26: 0.7691 - val_loss: 1.3588 - val_auc_26: 0.7545
Epoch 6/50
196038/196038 [==============================] - 17s 88us/sample - loss: 1.3014 - auc_26: 0.7733 - val_loss: 1.3592 - val_auc_26: 0.7544
Epoch 7/50


[I 2021-01-05 02:13:42,150] Trial 26 finished with value: 0.7544660736805222 and parameters: {'step': 2, 'no._units': 473, 'number': 2, 'best_weight': {0: 1.0, 1: 18.0}}. Best is trial 26 with value: 0.7544660736805222.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 20s 102us/sample - loss: 2.5206 - auc_27: 0.5263 - val_loss: 3.2936 - val_auc_27: 0.5000
Epoch 2/50
196038/196038 [==============================] - 20s 100us/sample - loss: 1.6702 - auc_27: 0.5004 - val_loss: 1.7223 - val_auc_27: 0.5000
Epoch 3/50
196038/196038 [==============================] - 19s 97us/sample - loss: 1.6203 - auc_27: 0.4996 - val_loss: 1.6669 - val_auc_27: 0.5000
Epoch 4/50
196038/196038 [==============================] - 19s 97us/sample - loss: 1.5988 - auc_27: 0.5012 - val_loss: 1.5864 - val_auc_27: 0.5000
Epoch 5/50
196038/196038 [==============================] - 19s 97us/sample - loss: 1.5911 - auc_27: 0.4975 - val_loss: 1.5838 - val_auc_27: 0.4999
Epoch 6/50
196038/196038 [==============================] - 19s 96us/sample - loss: 1.5860 - auc_27: 0.4953 - val_loss: 1.5847 - val_auc_27: 0.4995
Epoch 7/5

[I 2021-01-05 02:16:36,644] Trial 27 finished with value: 0.49948240875396827 and parameters: {'step': 2, 'no._units': 404, 'number': 3, 'best_weight': {0: 1.0, 1: 18.0}}. Best is trial 26 with value: 0.7544660736805222.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 18s 93us/sample - loss: 7.5081 - auc_28: 0.6215 - val_loss: 2.1105 - val_auc_28: 0.7221
Epoch 2/50
196038/196038 [==============================] - 17s 88us/sample - loss: 1.5435 - auc_28: 0.6982 - val_loss: 1.4231 - val_auc_28: 0.7448
Epoch 3/50
196038/196038 [==============================] - 17s 87us/sample - loss: 1.4283 - auc_28: 0.7264 - val_loss: 1.3942 - val_auc_28: 0.7488
Epoch 4/50
196038/196038 [==============================] - 17s 89us/sample - loss: 1.3711 - auc_28: 0.7469 - val_loss: 1.3686 - val_auc_28: 0.7503
Epoch 5/50
196038/196038 [==============================] - 17s 89us/sample - loss: 1.3323 - auc_28: 0.7626 - val_loss: 1.4335 - val_auc_28: 0.7523
Epoch 6/50
196038/196038 [==============================] - 17s 89us/sample - loss: 1.3084 - auc_28: 0.7725 - val_loss: 1.3565 - val_auc_28: 0.7531
Epoch 7/50


[I 2021-01-05 02:19:31,758] Trial 28 finished with value: 0.7532887320406003 and parameters: {'step': 3, 'no._units': 472, 'number': 2, 'best_weight': {0: 1.0, 1: 18.0}}. Best is trial 26 with value: 0.7544660736805222.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 26s 133us/sample - loss: 289.0927 - auc_29: 0.5034 - val_loss: 1.6162 - val_auc_29: 0.5000
Epoch 2/50
196038/196038 [==============================] - 25s 127us/sample - loss: 1.6112 - auc_29: 0.4975 - val_loss: 1.6613 - val_auc_29: 0.5000
Epoch 3/50
196038/196038 [==============================] - 25s 127us/sample - loss: 1.6025 - auc_29: 0.5005 - val_loss: 1.6734 - val_auc_29: 0.5000
Epoch 4/50
196038/196038 [==============================] - 25s 129us/sample - loss: 1.5988 - auc_29: 0.4989 - val_loss: 1.5976 - val_auc_29: 0.5000
Epoch 5/50
196038/196038 [==============================] - 25s 127us/sample - loss: 1.5901 - auc_29: 0.5015 - val_loss: 1.5991 - val_auc_29: 0.5000
Epoch 6/50
196038/196038 [==============================] - 25s 128us/sample - loss: 1.5861 - auc_29: 0.4983 - val_loss: 1.5964 - val_auc_29: 0.5000
Epo

[I 2021-01-05 02:24:08,196] Trial 29 finished with value: 0.5 and parameters: {'step': 2, 'no._units': 451, 'number': 6, 'best_weight': {0: 1.0, 1: 18.0}}. Best is trial 26 with value: 0.7544660736805222.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 20s 104us/sample - loss: 6.5605 - auc_30: 0.5123 - val_loss: 3.9739 - val_auc_30: 0.5000
Epoch 2/50
196038/196038 [==============================] - 20s 102us/sample - loss: 1.6855 - auc_30: 0.4966 - val_loss: 1.7709 - val_auc_30: 0.5000
Epoch 3/50
196038/196038 [==============================] - 19s 95us/sample - loss: 1.6260 - auc_30: 0.4979 - val_loss: 1.6345 - val_auc_30: 0.5000
Epoch 4/50
196038/196038 [==============================] - 20s 101us/sample - loss: 1.6105 - auc_30: 0.5040 - val_loss: 1.6259 - val_auc_30: 0.4984
Epoch 5/50
196038/196038 [==============================] - 20s 100us/sample - loss: 1.5871 - auc_30: 0.5182 - val_loss: 1.5737 - val_auc_30: 0.5435
Epoch 6/50
196038/196038 [==============================] - 19s 97us/sample - loss: 1.5563 - auc_30: 0.5644 - val_loss: 1.5564 - val_auc_30: 0.5746
Epoch 7

[I 2021-01-05 02:30:30,786] Trial 30 finished with value: 0.5958220994732153 and parameters: {'step': 2, 'no._units': 475, 'number': 3, 'best_weight': {0: 1.0, 1: 18.0}}. Best is trial 26 with value: 0.7544660736805222.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 18s 90us/sample - loss: 2.7600 - auc_31: 0.6083 - val_loss: 2.7814 - val_auc_31: 0.5018
Epoch 2/50
196038/196038 [==============================] - 16s 84us/sample - loss: 1.6341 - auc_31: 0.6370 - val_loss: 1.6775 - val_auc_31: 0.7274
Epoch 3/50
196038/196038 [==============================] - 17s 84us/sample - loss: 1.4123 - auc_31: 0.7286 - val_loss: 1.7439 - val_auc_31: 0.7393
Epoch 4/50
196038/196038 [==============================] - 17s 84us/sample - loss: 1.3706 - auc_31: 0.7494 - val_loss: 1.3722 - val_auc_31: 0.7521
Epoch 5/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.3254 - auc_31: 0.7646 - val_loss: 1.3645 - val_auc_31: 0.7510
Epoch 6/50
196038/196038 [==============================] - 17s 84us/sample - loss: 1.3004 - auc_31: 0.7750 - val_loss: 1.3761 - val_auc_31: 0.7520
Epoch 7/50


[I 2021-01-05 02:33:01,636] Trial 31 finished with value: 0.7511683631081711 and parameters: {'step': 3, 'no._units': 342, 'number': 2, 'best_weight': {0: 1.0, 1: 18.0}}. Best is trial 26 with value: 0.7544660736805222.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 17s 89us/sample - loss: 12.3027 - auc_32: 0.6237 - val_loss: 1.8743 - val_auc_32: 0.7179
Epoch 2/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.9375 - auc_32: 0.7054 - val_loss: 1.3904 - val_auc_32: 0.7406
Epoch 3/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.2689 - auc_32: 0.7259 - val_loss: 1.2337 - val_auc_32: 0.7462
Epoch 4/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.2193 - auc_32: 0.7469 - val_loss: 1.2194 - val_auc_32: 0.7535
Epoch 5/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.1863 - auc_32: 0.7625 - val_loss: 1.1992 - val_auc_32: 0.7551
Epoch 6/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.1583 - auc_32: 0.7740 - val_loss: 1.2162 - val_auc_32: 0.7538
Epoch 7/50

[I 2021-01-05 02:35:33,307] Trial 32 finished with value: 0.7551535308664359 and parameters: {'step': 3, 'no._units': 419, 'number': 2, 'best_weight': {0: 1.0, 1: 14.0}}. Best is trial 32 with value: 0.7551535308664359.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 18s 91us/sample - loss: 5.7413 - auc_33: 0.6482 - val_loss: 1.5792 - val_auc_33: 0.7356
Epoch 2/50
196038/196038 [==============================] - 17s 86us/sample - loss: 3.3577 - auc_33: 0.7081 - val_loss: 1.2717 - val_auc_33: 0.7335
Epoch 3/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.2723 - auc_33: 0.7359 - val_loss: 2.1895 - val_auc_33: 0.7495
Epoch 4/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.2180 - auc_33: 0.7487 - val_loss: 1.2501 - val_auc_33: 0.7524
Epoch 5/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.1952 - auc_33: 0.7614 - val_loss: 1.2714 - val_auc_33: 0.7541
Epoch 6/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.1702 - auc_33: 0.7689 - val_loss: 1.2600 - val_auc_33: 0.7528
Epoch 7/50


[I 2021-01-05 02:38:40,070] Trial 33 finished with value: 0.7526480758032611 and parameters: {'step': 3, 'no._units': 376, 'number': 2, 'best_weight': {0: 1.0, 1: 14.0}}. Best is trial 32 with value: 0.7551535308664359.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 18s 90us/sample - loss: 2.5395 - auc_34: 0.6073 - val_loss: 4.8404 - val_auc_34: 0.6859
Epoch 2/50
196038/196038 [==============================] - 16s 84us/sample - loss: 1.7541 - auc_34: 0.7166 - val_loss: 1.4124 - val_auc_34: 0.7503
Epoch 3/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.2176 - auc_34: 0.7461 - val_loss: 1.2738 - val_auc_34: 0.7547
Epoch 4/50
196038/196038 [==============================] - 17s 87us/sample - loss: 1.1828 - auc_34: 0.7619 - val_loss: 1.2448 - val_auc_34: 0.7560
Epoch 5/50
196038/196038 [==============================] - 17s 84us/sample - loss: 1.1641 - auc_34: 0.7711 - val_loss: 1.2459 - val_auc_34: 0.7559
Epoch 6/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.1549 - auc_34: 0.7748 - val_loss: 1.2415 - val_auc_34: 0.7561
Epoch 7/50


[I 2021-01-05 02:41:28,792] Trial 34 finished with value: 0.7560548910166114 and parameters: {'step': 2, 'no._units': 423, 'number': 2, 'best_weight': {0: 1.0, 1: 14.0}}. Best is trial 34 with value: 0.7560548910166114.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 20s 101us/sample - loss: 7.6501 - auc_35: 0.5156 - val_loss: 2.0448 - val_auc_35: 0.5001
Epoch 2/50
196038/196038 [==============================] - 19s 97us/sample - loss: 1.4925 - auc_35: 0.4991 - val_loss: 1.5309 - val_auc_35: 0.5001
Epoch 3/50
196038/196038 [==============================] - 19s 96us/sample - loss: 1.4476 - auc_35: 0.4994 - val_loss: 1.4098 - val_auc_35: 0.5000
Epoch 4/50
196038/196038 [==============================] - 19s 96us/sample - loss: 1.4245 - auc_35: 0.4964 - val_loss: 1.4095 - val_auc_35: 0.5002
Epoch 5/50
196038/196038 [==============================] - 19s 96us/sample - loss: 1.4152 - auc_35: 0.4971 - val_loss: 1.4116 - val_auc_35: 0.5012
Epoch 6/50
196038/196038 [==============================] - 19s 96us/sample - loss: 1.4100 - auc_35: 0.5031 - val_loss: 1.4098 - val_auc_35: 0.5014
Epoch 7/50

[I 2021-01-05 02:44:58,768] Trial 35 finished with value: 0.5015011767336891 and parameters: {'step': 2, 'no._units': 410, 'number': 3, 'best_weight': {0: 1.0, 1: 14.0}}. Best is trial 34 with value: 0.7560548910166114.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 17s 87us/sample - loss: 6.7985 - auc_36: 0.6170 - val_loss: 1.3506 - val_auc_36: 0.6902
Epoch 2/50
196038/196038 [==============================] - 16s 82us/sample - loss: 2.3482 - auc_36: 0.7056 - val_loss: 1.2469 - val_auc_36: 0.7375
Epoch 3/50
196038/196038 [==============================] - 16s 82us/sample - loss: 1.3134 - auc_36: 0.7398 - val_loss: 1.3057 - val_auc_36: 0.7462
Epoch 4/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.2107 - auc_36: 0.7529 - val_loss: 1.3155 - val_auc_36: 0.7487
Epoch 5/50
196038/196038 [==============================] - 16s 82us/sample - loss: 1.1875 - auc_36: 0.7600 - val_loss: 1.2885 - val_auc_36: 0.7490
Epoch 6/50
196038/196038 [==============================] - 16s 82us/sample - loss: 1.1793 - auc_36: 0.7625 - val_loss: 1.2850 - val_auc_36: 0.7492


[I 2021-01-05 02:46:38,051] Trial 36 finished with value: 0.7388634908085616 and parameters: {'step': 2, 'no._units': 328, 'number': 2, 'best_weight': {0: 1.0, 1: 14.0}}. Best is trial 34 with value: 0.7560548910166114.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 20s 100us/sample - loss: 6.1159 - auc_37: 0.5106 - val_loss: 2.4770 - val_auc_37: 0.5000
Epoch 2/50
196038/196038 [==============================] - 19s 95us/sample - loss: 1.4880 - auc_37: 0.4986 - val_loss: 1.7100 - val_auc_37: 0.5000
Epoch 3/50
196038/196038 [==============================] - 19s 95us/sample - loss: 1.4446 - auc_37: 0.4982 - val_loss: 1.4803 - val_auc_37: 0.5006
Epoch 4/50
196038/196038 [==============================] - 19s 95us/sample - loss: 1.3780 - auc_37: 0.6034 - val_loss: 1.2884 - val_auc_37: 0.7037
Epoch 5/50
196038/196038 [==============================] - 19s 95us/sample - loss: 1.2567 - auc_37: 0.7182 - val_loss: 1.2575 - val_auc_37: 0.7253
Epoch 6/50
196038/196038 [==============================] - 19s 95us/sample - loss: 1.2304 - auc_37: 0.7347 - val_loss: 1.2482 - val_auc_37: 0.7291
Epoch 7/50

[I 2021-01-05 02:50:43,462] Trial 37 finished with value: 0.73035522021293 and parameters: {'step': 2, 'no._units': 426, 'number': 3, 'best_weight': {0: 1.0, 1: 14.0}}. Best is trial 34 with value: 0.7560548910166114.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 18s 91us/sample - loss: 3.4908 - auc_38: 0.6232 - val_loss: 1.9634 - val_auc_38: 0.7019
Epoch 2/50
196038/196038 [==============================] - 17s 86us/sample - loss: 1.3192 - auc_38: 0.7019 - val_loss: 1.2816 - val_auc_38: 0.7276
Epoch 3/50
196038/196038 [==============================] - 17s 86us/sample - loss: 1.2259 - auc_38: 0.7413 - val_loss: 1.2511 - val_auc_38: 0.7439
Epoch 4/50
196038/196038 [==============================] - 17s 86us/sample - loss: 1.1881 - auc_38: 0.7593 - val_loss: 1.2225 - val_auc_38: 0.7500
Epoch 5/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.1706 - auc_38: 0.7674 - val_loss: 1.2199 - val_auc_38: 0.7518
Epoch 6/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.1621 - auc_38: 0.7713 - val_loss: 1.2218 - val_auc_38: 0.7514
Epoch 7/50


[I 2021-01-05 02:53:16,604] Trial 38 finished with value: 0.7519335310344103 and parameters: {'step': 2, 'no._units': 373, 'number': 2, 'best_weight': {0: 1.0, 1: 14.0}}. Best is trial 34 with value: 0.7560548910166114.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 18s 90us/sample - loss: 2.6114 - auc_39: 0.6398 - val_loss: 1.7491 - val_auc_39: 0.7216
Epoch 2/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.3285 - auc_39: 0.7034 - val_loss: 1.5527 - val_auc_39: 0.7364
Epoch 3/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.2558 - auc_39: 0.7339 - val_loss: 1.6339 - val_auc_39: 0.7469
Epoch 4/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.2086 - auc_39: 0.7509 - val_loss: 1.2471 - val_auc_39: 0.7494
Epoch 5/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.1816 - auc_39: 0.7629 - val_loss: 1.3187 - val_auc_39: 0.7538
Epoch 6/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.1618 - auc_39: 0.7716 - val_loss: 1.2422 - val_auc_39: 0.7539
Epoch 7/50


[I 2021-01-05 02:56:22,092] Trial 39 finished with value: 0.7541355961620114 and parameters: {'step': 3, 'no._units': 317, 'number': 2, 'best_weight': {0: 1.0, 1: 14.0}}. Best is trial 34 with value: 0.7560548910166114.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 19s 95us/sample - loss: 2.0046 - auc_40: 0.5088 - val_loss: 1.8229 - val_auc_40: 0.5000
Epoch 2/50
196038/196038 [==============================] - 18s 90us/sample - loss: 1.5301 - auc_40: 0.5007 - val_loss: 1.5047 - val_auc_40: 0.5000
Epoch 3/50
196038/196038 [==============================] - 17s 89us/sample - loss: 1.4601 - auc_40: 0.5021 - val_loss: 1.4105 - val_auc_40: 0.5000
Epoch 4/50
196038/196038 [==============================] - 17s 89us/sample - loss: 1.4383 - auc_40: 0.4985 - val_loss: 1.4098 - val_auc_40: 0.4999
Epoch 5/50
196038/196038 [==============================] - 17s 88us/sample - loss: 1.4261 - auc_40: 0.4979 - val_loss: 1.4137 - val_auc_40: 0.5000
Epoch 6/50
196038/196038 [==============================] - 17s 88us/sample - loss: 1.4177 - auc_40: 0.4996 - val_loss: 1.4109 - val_auc_40: 0.5000
Epoch 7/50


[I 2021-01-05 03:03:40,490] Trial 40 finished with value: 0.499968555436765 and parameters: {'step': 3, 'no._units': 301, 'number': 3, 'best_weight': {0: 1.0, 1: 14.0}}. Best is trial 34 with value: 0.7560548910166114.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 17s 86us/sample - loss: 2.3313 - auc_41: 0.6289 - val_loss: 4.5032 - val_auc_41: 0.6992
Epoch 2/50
196038/196038 [==============================] - 16s 82us/sample - loss: 1.3142 - auc_41: 0.7103 - val_loss: 1.3628 - val_auc_41: 0.7392
Epoch 3/50
196038/196038 [==============================] - 16s 82us/sample - loss: 1.3105 - auc_41: 0.7348 - val_loss: 1.2331 - val_auc_41: 0.7531
Epoch 4/50
196038/196038 [==============================] - 16s 82us/sample - loss: 1.2035 - auc_41: 0.7537 - val_loss: 1.2171 - val_auc_41: 0.7540
Epoch 5/50
196038/196038 [==============================] - 16s 81us/sample - loss: 1.1736 - auc_41: 0.7683 - val_loss: 1.2072 - val_auc_41: 0.7538
Epoch 6/50
196038/196038 [==============================] - 16s 82us/sample - loss: 1.1542 - auc_41: 0.7769 - val_loss: 1.2234 - val_auc_41: 0.7538
Epoch 7/50


[I 2021-01-05 03:06:07,165] Trial 41 finished with value: 0.7539029030165683 and parameters: {'step': 3, 'no._units': 325, 'number': 2, 'best_weight': {0: 1.0, 1: 14.0}}. Best is trial 34 with value: 0.7560548910166114.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 17s 89us/sample - loss: 1.9667 - auc_42: 0.6488 - val_loss: 3.2101 - val_auc_42: 0.7320
Epoch 2/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.3329 - auc_42: 0.7112 - val_loss: 1.6445 - val_auc_42: 0.7358
Epoch 3/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.2483 - auc_42: 0.7375 - val_loss: 1.3830 - val_auc_42: 0.7495
Epoch 4/50
196038/196038 [==============================] - 16s 84us/sample - loss: 1.2040 - auc_42: 0.7539 - val_loss: 1.2497 - val_auc_42: 0.7521
Epoch 5/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.1812 - auc_42: 0.7643 - val_loss: 1.2497 - val_auc_42: 0.7545
Epoch 6/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.1589 - auc_42: 0.7733 - val_loss: 1.2652 - val_auc_42: 0.7526
Epoch 7/50


[I 2021-01-05 03:08:36,820] Trial 42 finished with value: 0.7545677084082717 and parameters: {'step': 3, 'no._units': 269, 'number': 2, 'best_weight': {0: 1.0, 1: 14.0}}. Best is trial 34 with value: 0.7560548910166114.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 17s 87us/sample - loss: 1.6996 - auc_43: 0.6550 - val_loss: 1.7286 - val_auc_43: 0.7236
Epoch 2/50
196038/196038 [==============================] - 16s 82us/sample - loss: 1.2613 - auc_43: 0.7268 - val_loss: 1.2254 - val_auc_43: 0.7462
Epoch 3/50
196038/196038 [==============================] - 16s 82us/sample - loss: 1.2078 - auc_43: 0.7503 - val_loss: 1.2341 - val_auc_43: 0.7519
Epoch 4/50
196038/196038 [==============================] - 16s 82us/sample - loss: 1.1816 - auc_43: 0.7622 - val_loss: 1.2164 - val_auc_43: 0.7535
Epoch 5/50
196038/196038 [==============================] - 16s 82us/sample - loss: 1.1648 - auc_43: 0.7696 - val_loss: 1.2090 - val_auc_43: 0.7535
Epoch 6/50
196038/196038 [==============================] - 16s 82us/sample - loss: 1.1591 - auc_43: 0.7719 - val_loss: 1.2117 - val_auc_43: 0.7535
Epoch 7/50


[I 2021-01-05 03:11:03,900] Trial 43 finished with value: 0.753498503191505 and parameters: {'step': 2, 'no._units': 219, 'number': 2, 'best_weight': {0: 1.0, 1: 14.0}}. Best is trial 34 with value: 0.7560548910166114.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 17s 89us/sample - loss: 5.5974 - auc_44: 0.6326 - val_loss: 1.4351 - val_auc_44: 0.6841
Epoch 2/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.2964 - auc_44: 0.7119 - val_loss: 1.7299 - val_auc_44: 0.7357
Epoch 3/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.2603 - auc_44: 0.7291 - val_loss: 1.2292 - val_auc_44: 0.7472
Epoch 4/50
196038/196038 [==============================] - 16s 84us/sample - loss: 1.2181 - auc_44: 0.7479 - val_loss: 1.2453 - val_auc_44: 0.7473
Epoch 5/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.1855 - auc_44: 0.7602 - val_loss: 1.2417 - val_auc_44: 0.7480
Epoch 6/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.1638 - auc_44: 0.7697 - val_loss: 1.2511 - val_auc_44: 0.7479
Epoch 7/50


[I 2021-01-05 03:13:00,886] Trial 44 finished with value: 0.7472025484484387 and parameters: {'step': 3, 'no._units': 257, 'number': 2, 'best_weight': {0: 1.0, 1: 14.0}}. Best is trial 34 with value: 0.7560548910166114.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 17s 87us/sample - loss: 2.0423 - auc_45: 0.6686 - val_loss: 1.2831 - val_auc_45: 0.7270
Epoch 2/50
196038/196038 [==============================] - 16s 82us/sample - loss: 1.3062 - auc_45: 0.7208 - val_loss: 1.2220 - val_auc_45: 0.7433
Epoch 3/50
196038/196038 [==============================] - 16s 82us/sample - loss: 1.2273 - auc_45: 0.7417 - val_loss: 1.2673 - val_auc_45: 0.7511
Epoch 4/50
196038/196038 [==============================] - 16s 82us/sample - loss: 1.1943 - auc_45: 0.7573 - val_loss: 1.2561 - val_auc_45: 0.7541
Epoch 5/50
196038/196038 [==============================] - 16s 82us/sample - loss: 1.1715 - auc_45: 0.7672 - val_loss: 1.2073 - val_auc_45: 0.7541
Epoch 6/50
196038/196038 [==============================] - 16s 82us/sample - loss: 1.1548 - auc_45: 0.7744 - val_loss: 1.2081 - val_auc_45: 0.7537
Epoch 7/50


[I 2021-01-05 03:15:28,139] Trial 45 finished with value: 0.7541960985189127 and parameters: {'step': 3, 'no._units': 169, 'number': 2, 'best_weight': {0: 1.0, 1: 14.0}}. Best is trial 34 with value: 0.7560548910166114.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 17s 88us/sample - loss: 1.5381 - auc_46: 0.6800 - val_loss: 1.8297 - val_auc_46: 0.7410
Epoch 2/50
196038/196038 [==============================] - 16s 83us/sample - loss: 2.0011 - auc_46: 0.7366 - val_loss: 1.3108 - val_auc_46: 0.7464
Epoch 3/50
196038/196038 [==============================] - 16s 84us/sample - loss: 1.1608 - auc_46: 0.7543 - val_loss: 1.1948 - val_auc_46: 0.7527
Epoch 4/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.1373 - auc_46: 0.7637 - val_loss: 1.1996 - val_auc_46: 0.7531
Epoch 5/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.1274 - auc_46: 0.7682 - val_loss: 1.2046 - val_auc_46: 0.7541
Epoch 6/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.1233 - auc_46: 0.7700 - val_loss: 1.2065 - val_auc_46: 0.7542
Epoch 7/50


[I 2021-01-05 03:17:25,062] Trial 46 finished with value: 0.7527851968408268 and parameters: {'step': 2, 'no._units': 133, 'number': 2, 'best_weight': {0: 1.0, 1: 13.0}}. Best is trial 34 with value: 0.7560548910166114.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 18s 94us/sample - loss: 1.8444 - auc_47: 0.5709 - val_loss: 1.8190 - val_auc_47: 0.5000
Epoch 2/50
196038/196038 [==============================] - 17s 89us/sample - loss: 1.5843 - auc_47: 0.4990 - val_loss: 1.5778 - val_auc_47: 0.5000
Epoch 3/50
196038/196038 [==============================] - 17s 89us/sample - loss: 1.5619 - auc_47: 0.4996 - val_loss: 1.5469 - val_auc_47: 0.5000
Epoch 4/50
196038/196038 [==============================] - 17s 89us/sample - loss: 1.5512 - auc_47: 0.4996 - val_loss: 1.5468 - val_auc_47: 0.5006
Epoch 5/50
196038/196038 [==============================] - 17s 89us/sample - loss: 1.5476 - auc_47: 0.4975 - val_loss: 1.5466 - val_auc_47: 0.5010
Epoch 6/50
196038/196038 [==============================] - 17s 89us/sample - loss: 1.5434 - auc_47: 0.4987 - val_loss: 1.5434 - val_auc_47: 0.5012
Epoch 7/50


[I 2021-01-05 03:20:41,926] Trial 47 finished with value: 0.5013786634013858 and parameters: {'step': 2, 'no._units': 169, 'number': 3, 'best_weight': {0: 1.0, 1: 17.0}}. Best is trial 34 with value: 0.7560548910166114.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 17s 85us/sample - loss: 1.7097 - auc_48: 0.6617 - val_loss: 1.8028 - val_auc_48: 0.7254
Epoch 2/50
196038/196038 [==============================] - 16s 81us/sample - loss: 1.3167 - auc_48: 0.7232 - val_loss: 1.5523 - val_auc_48: 0.7437
Epoch 3/50
196038/196038 [==============================] - 16s 80us/sample - loss: 1.2666 - auc_48: 0.7431 - val_loss: 1.3944 - val_auc_48: 0.7471
Epoch 4/50
196038/196038 [==============================] - 16s 80us/sample - loss: 1.2327 - auc_48: 0.7585 - val_loss: 1.2690 - val_auc_48: 0.7524
Epoch 5/50
196038/196038 [==============================] - 16s 80us/sample - loss: 1.2071 - auc_48: 0.7698 - val_loss: 1.2567 - val_auc_48: 0.7535
Epoch 6/50
196038/196038 [==============================] - 16s 80us/sample - loss: 1.1904 - auc_48: 0.7769 - val_loss: 1.2795 - val_auc_48: 0.7526
Epoch 7/50


[I 2021-01-05 03:23:05,694] Trial 48 finished with value: 0.7535045658113768 and parameters: {'step': 3, 'no._units': 182, 'number': 2, 'best_weight': {0: 1.0, 1: 15.0}}. Best is trial 34 with value: 0.7560548910166114.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 17s 89us/sample - loss: 1.5389 - auc_49: 0.6746 - val_loss: 1.4092 - val_auc_49: 0.7389
Epoch 2/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.2659 - auc_49: 0.7263 - val_loss: 1.2620 - val_auc_49: 0.7453
Epoch 3/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.2205 - auc_49: 0.7455 - val_loss: 1.6851 - val_auc_49: 0.7400
Epoch 4/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.1900 - auc_49: 0.7594 - val_loss: 1.2085 - val_auc_49: 0.7559
Epoch 5/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.1641 - auc_49: 0.7696 - val_loss: 1.2206 - val_auc_49: 0.7562
Epoch 6/50
196038/196038 [==============================] - 16s 83us/sample - loss: 1.1489 - auc_49: 0.7761 - val_loss: 1.2157 - val_auc_49: 0.7561
Epoch 7/50


[I 2021-01-05 03:25:18,202] Trial 49 finished with value: 0.7558950844097945 and parameters: {'step': 3, 'no._units': 142, 'number': 2, 'best_weight': {0: 1.0, 1: 14.0}}. Best is trial 34 with value: 0.7560548910166114.


In [19]:
print('The best roc_auc_score for the study is: ',best_study_score)

The best roc_auc_score for the study is:  0.7560548910166114


In [20]:
print('The best study parameters for the classifier are: ',best_study_params)

The best study parameters for the classifier are:  {'step': 2, 'no._units': 423, 'number': 2, 'best_weight': {0: 1.0, 1: 14.0}}


In [21]:
# loading the best Neural Network
Best_Model_selu = keras.models.load_model('Best_model_Selu_eq_Learn.h5')

In [22]:
# Printing its summary
Best_Model_selu.summary()

Model: "sequential_34"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_134 (Dense)            (None, 423)               43992     
_________________________________________________________________
dense_135 (Dense)            (None, 423)               179352    
_________________________________________________________________
dense_136 (Dense)            (None, 1)                 424       
Total params: 223,768
Trainable params: 223,768
Non-trainable params: 0
_________________________________________________________________


In [23]:
# Defining the function to calculate the roc_auc score for the feature sets
def cal_roc_auc(X, y, cls, f_set, t_set, model_name):
    ''' Calculates the roc auc score using the best study parameters 
        f_set : String: specifies 'full feature', 'Reduced feature'
        t_set: String: specifies 'training', 'test'
        model_name: String: specifies Name of the model '''
        
    y_pred = cls.predict_proba(X)
    print('The roc_auc_score for the {} {} set using the best {} classifier is '.format(f_set,t_set,model_name),roc_auc_score(y,y_pred))

In [24]:
# Calculating the Reduced feature training set roc_auc score using the pretrained Neural Network having 
# best study parameters & weights
cal_roc_auc(X_train_scaled, y_train, Best_Model_selu, 'Reduced feature', 'training', 'Dense Neural Network using SELU activation')

The roc_auc_score for the Reduced feature training set using the best Dense Neural Network using SELU activation classifier is  0.7735250280260082


In [25]:
# Calculating the Reduced feature test set roc_auc score using the pretrained Neural Network having 
# best study parameters & weights
cal_roc_auc(X_test_scaled, y_test, Best_Model_selu, 'Reduced feature', 'test', 'Dense Neural Network using SELU activation')

The roc_auc_score for the Reduced feature test set using the best Dense Neural Network using SELU activation classifier is  0.757445351657831


## Observation(s):
### 1) Though the tuned Selu dense neural classifier has test set roc_auc score of 0.757445351657831, with almost no overfitting, certain optima trials have ended with roc_auc score of .5, corresponding to random guessing. It may be that dataset is allergic to Nadam optimizer or Selu sctivation function. thus let's discard this model for this dataset.

## Model_11.1: Keras Dense Model with Equal Neurons/layer, SELU Activation, RMS Prop Optimizer & Tuned learning rate.

In [33]:
# Defining best roc_auc score
best_roc_auc = 0

In [34]:
# defining the objective function
def objective_wrappper_keras_eq_RMS(X_tr, y_tr, X_val, y_val, cl_w):
    '''
    Optimizes Keras ,neural net with equal no. of neurons in each layer , best parameters 
    on the given training set: X_tr, y_tr using validation set: X_val,y_val & class weights: cl_w
    
    '''

    def objective(trial):
        s = trial.suggest_int('step',2,6)
        def exponential_decay_fn(epoch):
            return 0.01 * 0.1**(epoch /s )
              
        # building model & searching for the best no. of neural units per layer.
        no_units = trial.suggest_int('no._units',100,600)# No. of Neurons in each layer
        model = keras.models.Sequential()
        model.add(keras.layers.InputLayer(input_shape=X_tr.shape[1:]))
        #model.add(keras.layers.BatchNormalization())
        for i in range(trial.suggest_int('number',2,6)):
            model.add(keras.layers.Dense(units=no_units,activation='selu',
                                     kernel_initializer='lecun_normal')) # For self normalization
        model.add(keras.layers.Dense(units=1,activation='sigmoid'))
    
        # Searching for the best learning rate.
        #lr = trial.suggest_float("lr", 1e-5, 1e-1, log=True) 
    
        # Compiling the model
        model.compile(loss="binary_crossentropy", optimizer=keras.optimizers.RMSprop(),
                      metrics=[tf.keras.metrics.AUC(curve="ROC")])
        
        # Defining the Callbacks
        checkpoint_cb = keras.callbacks.ModelCheckpoint('best_model.h5',save_best_only=True) # 1st Callback
        early_stopping_cb = keras.callbacks.EarlyStopping(patience=4,restore_best_weights=True) # 2nd Callback, Stop if validation score doen't improve for
        # 4 epochs        
        lr_scheduler_cb = keras.callbacks.LearningRateScheduler(exponential_decay_fn)# 3rd Callback
        #tensorboard_cb = keras.callbacks.TensorBoard(get_run_logdir()) 
        
        #Fitting the model
        model.fit(X_tr, y_tr, epochs=50, validation_data=(X_val,y_val), batch_size=32,
             class_weight=trial.suggest_categorical('best_weight',cl_w),
             callbacks=[checkpoint_cb, early_stopping_cb, lr_scheduler_cb])#,tensorboard_cb])
    
        # Loading the best model 
        model_best = keras.models.load_model('best_model.h5')
    
        # Computing the roc_auc_score for the validation set
        roc_auc_1 = roc_auc_score(y_val,model_best.predict_proba(X_val))
        global best_roc_auc
        # Updating best_roc_auc score
        if roc_auc_1 > best_roc_auc:
            best_roc_auc = roc_auc_1
            model_best.save('Best_model_Selu_eq_Learn_RMS.h5')
        
        return roc_auc_1
    
    return objective

In [35]:
# Defining the evaluation function for study's best parameters
def study_best_score_params(X_tr, y_tr, X_v, y_v, obj_func, cl_w, n_trials=100):
    ''' Computes the best hyper parameters of the Neural net using the Training set(X_tr,y_tr) & Validation set
    (X_v,y_v), class weights cl_w & 
    returns Optuna's study's best score & clasifier parameters'''
    
    study = optuna.create_study(direction='maximize')
    study.optimize(obj_func(X_tr, y_tr, X_v, y_v, cl_w),n_trials)
    best_score = study.best_value
    best_params = study.best_params
    return (best_score,best_params)

In [36]:
# Extracting the best model parameters and best study score
best_study_score,best_study_params = study_best_score_params(X_train_1, y_train_1, X_val, y_val, objective_wrappper_keras_eq_RMS, cl_weight,
                                                            n_trials=10)

[I 2021-01-05 20:56:57,233] A new study created in memory with name: no-name-d7c2a11f-f359-46c1-89c6-60b330e460f7


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 16s 84us/sample - loss: 9.4790 - auc_2: 0.5295 - val_loss: 29.4142 - val_auc_2: 0.5079
Epoch 2/50
196038/196038 [==============================] - 15s 79us/sample - loss: 3.5154 - auc_2: 0.5101 - val_loss: 3.8783 - val_auc_2: 0.5033
Epoch 3/50
196038/196038 [==============================] - 16s 79us/sample - loss: 2.2828 - auc_2: 0.5176 - val_loss: 3.3285 - val_auc_2: 0.5792
Epoch 4/50
196038/196038 [==============================] - 16s 80us/sample - loss: 1.8605 - auc_2: 0.5406 - val_loss: 1.9503 - val_auc_2: 0.5669
Epoch 5/50
196038/196038 [==============================] - 16s 80us/sample - loss: 1.6091 - auc_2: 0.6301 - val_loss: 1.4378 - val_auc_2: 0.7199
Epoch 6/50
196038/196038 [==============================] - 16s 80us/sample - loss: 1.4431 - auc_2: 0.6926 - val_loss: 1.4145 - val_auc_2: 0.7154
Epoch 7/50
196038/1960

[I 2021-01-05 21:00:39,010] Trial 0 finished with value: 0.7336060679157983 and parameters: {'step': 4, 'no._units': 394, 'number': 3, 'best_weight': {0: 1.0, 1: 15.0}}. Best is trial 0 with value: 0.7336060679157983.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 19s 95us/sample - loss: 2.3205 - auc_3: 0.5028 - val_loss: 1.6852 - val_auc_3: 0.5000
Epoch 2/50
196038/196038 [==============================] - 18s 89us/sample - loss: 1.5733 - auc_3: 0.4973 - val_loss: 2.3404 - val_auc_3: 0.5000
Epoch 3/50
196038/196038 [==============================] - 17s 89us/sample - loss: 1.4606 - auc_3: 0.4960 - val_loss: 1.8156 - val_auc_3: 0.5000
Epoch 4/50
196038/196038 [==============================] - 18s 89us/sample - loss: 1.3959 - auc_3: 0.5003 - val_loss: 1.8611 - val_auc_3: 0.5000
Epoch 5/50
196038/196038 [==============================] - 18s 90us/sample - loss: 1.3637 - auc_3: 0.4980 - val_loss: 1.4810 - val_auc_3: 0.5000
Epoch 6/50
196038/196038 [==============================] - 18s 90us/sample - loss: 1.3425 - auc_3: 0.4988 - val_loss: 1.3922 - val_auc_3: 0.5000
Epoch 7/50
196038/19603

[I 2021-01-05 21:07:28,566] Trial 1 finished with value: 0.5 and parameters: {'step': 6, 'no._units': 152, 'number': 6, 'best_weight': {0: 1.0, 1: 12.0}}. Best is trial 0 with value: 0.7336060679157983.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 18s 91us/sample - loss: 17.6369 - auc_4: 0.5322 - val_loss: 7.5313 - val_auc_4: 0.4847
Epoch 2/50
196038/196038 [==============================] - 17s 86us/sample - loss: 4.9082 - auc_4: 0.5134 - val_loss: 3.0246 - val_auc_4: 0.5151
Epoch 3/50
196038/196038 [==============================] - 17s 86us/sample - loss: 3.0655 - auc_4: 0.5074 - val_loss: 3.7671 - val_auc_4: 0.5054
Epoch 4/50
196038/196038 [==============================] - 17s 87us/sample - loss: 2.3253 - auc_4: 0.5201 - val_loss: 4.4731 - val_auc_4: 0.5338
Epoch 5/50
196038/196038 [==============================] - 17s 86us/sample - loss: 1.9895 - auc_4: 0.5543 - val_loss: 2.1131 - val_auc_4: 0.5969
Epoch 6/50
196038/196038 [==============================] - 17s 86us/sample - loss: 1.8196 - auc_4: 0.6573 - val_loss: 1.9477 - val_auc_4: 0.7206
Epoch 7/50
196038/1960

[I 2021-01-05 21:12:01,644] Trial 2 finished with value: 0.7229389304700925 and parameters: {'step': 4, 'no._units': 571, 'number': 3, 'best_weight': {0: 1.0, 1: 17.0}}. Best is trial 0 with value: 0.7336060679157983.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 17s 84us/sample - loss: 3.1615 - auc_5: 0.5409 - val_loss: 5.4203 - val_auc_5: 0.5001
Epoch 2/50
196038/196038 [==============================] - 16s 80us/sample - loss: 1.9313 - auc_5: 0.5107 - val_loss: 2.4129 - val_auc_5: 0.5001
Epoch 3/50
196038/196038 [==============================] - 16s 79us/sample - loss: 1.5300 - auc_5: 0.4997 - val_loss: 1.7107 - val_auc_5: 0.5000
Epoch 4/50
196038/196038 [==============================] - 16s 80us/sample - loss: 1.4892 - auc_5: 0.5202 - val_loss: 2.4036 - val_auc_5: 0.5205
Epoch 5/50
196038/196038 [==============================] - 16s 80us/sample - loss: 1.3490 - auc_5: 0.6248 - val_loss: 2.2201 - val_auc_5: 0.6778
Epoch 6/50
196038/196038 [==============================] - 16s 79us/sample - loss: 1.3416 - auc_5: 0.6844 - val_loss: 2.8827 - val_auc_5: 0.6434
Epoch 7/50
196038/19603

[I 2021-01-05 21:13:53,404] Trial 3 finished with value: 0.5 and parameters: {'step': 5, 'no._units': 261, 'number': 3, 'best_weight': {0: 1.0, 1: 11.38697}}. Best is trial 0 with value: 0.7336060679157983.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 19s 98us/sample - loss: 7.7891 - auc_6: 0.5042 - val_loss: 1.8127 - val_auc_6: 0.5000
Epoch 2/50
196038/196038 [==============================] - 18s 94us/sample - loss: 2.6246 - auc_6: 0.5033 - val_loss: 2.7687 - val_auc_6: 0.5000
Epoch 3/50
196038/196038 [==============================] - 18s 94us/sample - loss: 1.8240 - auc_6: 0.4993 - val_loss: 1.3950 - val_auc_6: 0.5000
Epoch 4/50
196038/196038 [==============================] - 18s 94us/sample - loss: 1.4945 - auc_6: 0.5013 - val_loss: 1.9714 - val_auc_6: 0.5000
Epoch 5/50
196038/196038 [==============================] - 18s 94us/sample - loss: 1.3731 - auc_6: 0.5005 - val_loss: 1.8560 - val_auc_6: 0.5000
Epoch 6/50
196038/196038 [==============================] - 18s 94us/sample - loss: 1.3221 - auc_6: 0.4998 - val_loss: 1.3168 - val_auc_6: 0.5000
Epoch 7/50
196038/19603

[I 2021-01-05 21:18:50,755] Trial 4 finished with value: 0.5 and parameters: {'step': 4, 'no._units': 535, 'number': 4, 'best_weight': {0: 1.0, 1: 11.38697}}. Best is trial 0 with value: 0.7336060679157983.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 22s 110us/sample - loss: 52.5843 - auc_7: 0.4966 - val_loss: 5.8291 - val_auc_7: 0.5000
Epoch 2/50
196038/196038 [==============================] - 21s 105us/sample - loss: 2.8726 - auc_7: 0.4994 - val_loss: 3.1100 - val_auc_7: 0.5000
Epoch 3/50
196038/196038 [==============================] - 20s 104us/sample - loss: 1.9966 - auc_7: 0.4983 - val_loss: 2.3722 - val_auc_7: 0.5000
Epoch 4/50
196038/196038 [==============================] - 21s 105us/sample - loss: 1.6463 - auc_7: 0.5015 - val_loss: 1.5372 - val_auc_7: 0.5000
Epoch 5/50
196038/196038 [==============================] - 20s 103us/sample - loss: 1.5151 - auc_7: 0.4999 - val_loss: 1.7363 - val_auc_7: 0.5000
Epoch 6/50
196038/196038 [==============================] - 20s 102us/sample - loss: 1.4630 - auc_7: 0.4961 - val_loss: 1.5724 - val_auc_7: 0.5000
Epoch 7/50
19603

[I 2021-01-05 21:25:16,307] Trial 5 finished with value: 0.5 and parameters: {'step': 4, 'no._units': 525, 'number': 6, 'best_weight': {0: 1.0, 1: 14.0}}. Best is trial 0 with value: 0.7336060679157983.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 15s 79us/sample - loss: 34.8352 - auc_8: 0.5822 - val_loss: 96.2213 - val_auc_8: 0.5542
Epoch 2/50
196038/196038 [==============================] - 15s 75us/sample - loss: 40.2028 - auc_8: 0.6019 - val_loss: 53.8549 - val_auc_8: 0.5688
Epoch 3/50
196038/196038 [==============================] - 15s 75us/sample - loss: 26.0023 - auc_8: 0.6185 - val_loss: 15.6750 - val_auc_8: 0.6403
Epoch 4/50
196038/196038 [==============================] - 15s 75us/sample - loss: 13.3767 - auc_8: 0.6415 - val_loss: 15.2716 - val_auc_8: 0.6309
Epoch 5/50
196038/196038 [==============================] - 15s 75us/sample - loss: 9.2518 - auc_8: 0.6541 - val_loss: 12.3198 - val_auc_8: 0.6504
Epoch 6/50
196038/196038 [==============================] - 15s 75us/sample - loss: 7.5270 - auc_8: 0.6697 - val_loss: 6.6917 - val_auc_8: 0.6451
Epoch 7/50
196

[I 2021-01-05 21:27:59,846] Trial 6 finished with value: 0.6541726259912045 and parameters: {'step': 3, 'no._units': 404, 'number': 2, 'best_weight': {0: 1.0, 1: 14.0}}. Best is trial 0 with value: 0.7336060679157983.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 20s 104us/sample - loss: 16.9822 - auc_9: 0.5016 - val_loss: 6.7261 - val_auc_9: 0.5007
Epoch 2/50
196038/196038 [==============================] - 19s 99us/sample - loss: 1.9444 - auc_9: 0.5022 - val_loss: 4.2228 - val_auc_9: 0.5000
Epoch 3/50
196038/196038 [==============================] - 19s 99us/sample - loss: 1.4955 - auc_9: 0.5014 - val_loss: 1.9381 - val_auc_9: 0.5000
Epoch 4/50
196038/196038 [==============================] - 19s 99us/sample - loss: 1.3753 - auc_9: 0.4996 - val_loss: 1.4625 - val_auc_9: 0.5000
Epoch 5/50
196038/196038 [==============================] - 19s 99us/sample - loss: 1.3350 - auc_9: 0.5003 - val_loss: 1.4790 - val_auc_9: 0.5000
Epoch 6/50
196038/196038 [==============================] - 19s 99us/sample - loss: 1.3200 - auc_9: 0.5000 - val_loss: 1.3364 - val_auc_9: 0.5000
Epoch 7/50
196038/196

[I 2021-01-05 21:31:54,949] Trial 7 finished with value: 0.5 and parameters: {'step': 3, 'no._units': 387, 'number': 6, 'best_weight': {0: 1.0, 1: 12.0}}. Best is trial 0 with value: 0.7336060679157983.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 16s 82us/sample - loss: 2.9939 - auc_10: 0.5346 - val_loss: 2.3453 - val_auc_10: 0.4996
Epoch 2/50
196038/196038 [==============================] - 15s 76us/sample - loss: 1.7515 - auc_10: 0.5547 - val_loss: 2.0175 - val_auc_10: 0.5090
Epoch 3/50
196038/196038 [==============================] - 15s 77us/sample - loss: 1.4452 - auc_10: 0.6886 - val_loss: 1.8676 - val_auc_10: 0.7176
Epoch 4/50
196038/196038 [==============================] - 15s 76us/sample - loss: 1.4377 - auc_10: 0.7251 - val_loss: 1.4935 - val_auc_10: 0.7145
Epoch 5/50
196038/196038 [==============================] - 15s 76us/sample - loss: 1.4026 - auc_10: 0.7327 - val_loss: 1.4137 - val_auc_10: 0.7254
Epoch 6/50
196038/196038 [==============================] - 15s 76us/sample - loss: 1.3781 - auc_10: 0.7358 - val_loss: 1.4543 - val_auc_10: 0.7228
Epoch 7/50


[I 2021-01-05 21:34:42,042] Trial 8 finished with value: 0.7253752029908114 and parameters: {'step': 2, 'no._units': 202, 'number': 3, 'best_weight': {0: 1.0, 1: 16.0}}. Best is trial 0 with value: 0.7336060679157983.


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
196038/196038 [==============================] - 16s 82us/sample - loss: 3.1882 - auc_11: 0.5377 - val_loss: 3.5060 - val_auc_11: 0.5001
Epoch 2/50
196038/196038 [==============================] - 15s 76us/sample - loss: 2.1123 - auc_11: 0.5141 - val_loss: 4.4321 - val_auc_11: 0.5000
Epoch 3/50
196038/196038 [==============================] - 15s 77us/sample - loss: 1.7539 - auc_11: 0.5043 - val_loss: 2.5915 - val_auc_11: 0.5000
Epoch 4/50
196038/196038 [==============================] - 15s 77us/sample - loss: 1.7372 - auc_11: 0.5322 - val_loss: 2.1206 - val_auc_11: 0.4988
Epoch 5/50
196038/196038 [==============================] - 15s 77us/sample - loss: 1.6437 - auc_11: 0.5787 - val_loss: 1.4811 - val_auc_11: 0.6451
Epoch 6/50
196038/196038 [==============================] - 15s 77us/sample - loss: 1.5468 - auc_11: 0.6429 - val_loss: 1.4571 - val_auc_11: 0.6852
Epoch 7/50


[I 2021-01-05 21:37:15,059] Trial 9 finished with value: 0.6867356384528662 and parameters: {'step': 6, 'no._units': 215, 'number': 3, 'best_weight': {0: 1.0, 1: 15.0}}. Best is trial 0 with value: 0.7336060679157983.


## Observation(s):
### 1) Similar to the previous case, certain optima trials have again ended with roc_auc score of .5. It is almost surely now that SELU activation is not the right activation function for this dataset.

## Model_11.2: Keras Dense Model with Equal Neurons/layer, SELU Activation, Nesterov Optimizer & Tuned learning rate.

In [18]:
# Defining best roc_auc score
best_roc_auc = 0

In [19]:
# defining the objective function
def objective_wrappper_keras_eq_Nesterov(X_tr, y_tr, X_val, y_val, cl_w):
    '''
    Optimizes Keras ,neural net with equal no. of neurons in each layer , best parameters 
    on the given training set: X_tr, y_tr using validation set: X_val,y_val & class weights: cl_w
    
    '''

    def objective(trial):
        s = trial.suggest_int('step',2,6)
        def exponential_decay_fn(epoch):
            return 0.01 * 0.1**(epoch /s )
              
        # building model & searching for the best no. of neural units per layer.
        no_units = trial.suggest_int('no._units',100,600)# No. of Neurons in each layer
        model = keras.models.Sequential()
        model.add(keras.layers.InputLayer(input_shape=X_tr.shape[1:]))
        #model.add(keras.layers.BatchNormalization())
        for i in range(trial.suggest_int('number',2,6)):
            model.add(keras.layers.Dense(units=no_units,activation='selu',
                                     kernel_initializer='lecun_normal')) # For self normalization
        model.add(keras.layers.Dense(units=1,activation='sigmoid'))
    
        # Searching for the best learning rate.
        #lr = trial.suggest_float("lr", 1e-5, 1e-1, log=True) 
    
        # Compiling the model
        model.compile(loss="binary_crossentropy", optimizer=keras.optimizers.SGD(momentum=.9,nesterov=True),
                      metrics=[tf.keras.metrics.AUC(curve="ROC")])
        
        # Defining the Callbacks
        checkpoint_cb = keras.callbacks.ModelCheckpoint('best_model.h5',save_best_only=True) # 1st Callback
        early_stopping_cb = keras.callbacks.EarlyStopping(patience=4,restore_best_weights=True) # 2nd Callback, Stop if validation score doen't improve for
        # 4 epochs        
        lr_scheduler_cb = keras.callbacks.LearningRateScheduler(exponential_decay_fn)# 3rd Callback
        #tensorboard_cb = keras.callbacks.TensorBoard(get_run_logdir()) 
        
        #Fitting the model
        model.fit(X_tr, y_tr, epochs=50, validation_data=(X_val,y_val), batch_size=32,
             class_weight=trial.suggest_categorical('best_weight',cl_w),
             callbacks=[checkpoint_cb, early_stopping_cb, lr_scheduler_cb])#,tensorboard_cb])
    
        # Loading the best model 
        model_best = keras.models.load_model('best_model.h5')
    
        # Computing the roc_auc_score for the validation set
        roc_auc_1 = roc_auc_score(y_val,model_best.predict_proba(X_val))
        global best_roc_auc
        # Updating best_roc_auc score
        if roc_auc_1 > best_roc_auc:
            best_roc_auc = roc_auc_1
            model_best.save('Best_model_Selu_eq_Learn_Nesterov.h5')
        
        return roc_auc_1
    
    return objective

In [20]:
# Defining the evaluation function for study's best parameters
def study_best_score_params(X_tr, y_tr, X_v, y_v, obj_func, cl_w, n_trials=100):
    ''' Computes the best hyper parameters of the Neural net using the Training set(X_tr,y_tr) & Validation set
    (X_v,y_v), class weights cl_w & 
    returns Optuna's study's best score & clasifier parameters'''
    
    study = optuna.create_study(direction='maximize')
    study.optimize(obj_func(X_tr, y_tr, X_v, y_v, cl_w),n_trials)
    best_score = study.best_value
    best_params = study.best_params
    return (best_score,best_params)

In [21]:
# Extracting the best model parameters and best study score
best_study_score,best_study_params = study_best_score_params(X_train_1, y_train_1, X_val, y_val, objective_wrappper_keras_eq_Nesterov, cl_weight,
                                                            n_trials=10)

[I 2021-01-07 21:57:57,299] A new study created in memory with name: no-name-690ac7a3-e37b-4305-9089-64ed7f5bb558


  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 196038 samples, validate on 34595 samples
Epoch 1/50
    32/196038 [..............................] - ETA: 1:27:21 - loss: 6.4671 - auc: 0.4349WARNING:tensorflow:Can save best model only with val_loss available, skipping.


[W 2021-01-07 21:57:59,200] Trial 0 failed because of the following error: InvalidArgumentError()
Traceback (most recent call last):
  File "C:\Users\panka\anaconda3\lib\site-packages\optuna\study.py", line 778, in _run_trial
    result = func(trial)
  File "<ipython-input-19-fcb6aa84e5fb>", line 41, in objective
    callbacks=[checkpoint_cb, early_stopping_cb, lr_scheduler_cb])#,tensorboard_cb])
  File "C:\Users\panka\anaconda3\lib\site-packages\tensorflow_core\python\keras\engine\training.py", line 819, in fit
    use_multiprocessing=use_multiprocessing)
  File "C:\Users\panka\anaconda3\lib\site-packages\tensorflow_core\python\keras\engine\training_v2.py", line 342, in fit
    total_epochs=epochs)
  File "C:\Users\panka\anaconda3\lib\site-packages\tensorflow_core\python\keras\engine\training_v2.py", line 128, in run_one_epoch
    batch_outs = execution_function(iterator)
  File "C:\Users\panka\anaconda3\lib\site-packages\tensorflow_core\python\keras\engine\training_v2_utils.py", line

InvalidArgumentError: 2 root error(s) found.
  (0) Invalid argument:  assertion failed: [predictions must be >= 0] [Condition x >= y did not hold element-wise:] [x (sequential/dense_6/Sigmoid:0) = ] [[nan][nan][nan]...] [y (metrics/auc/Cast_2/x:0) = ] [0]
	 [[{{node metrics/auc/assert_greater_equal/Assert/AssertGuard/else/_24/Assert}}]]
	 [[metrics/auc/assert_less_equal/Assert/AssertGuard/pivot_f/_36/_71]]
  (1) Invalid argument:  assertion failed: [predictions must be >= 0] [Condition x >= y did not hold element-wise:] [x (sequential/dense_6/Sigmoid:0) = ] [[nan][nan][nan]...] [y (metrics/auc/Cast_2/x:0) = ] [0]
	 [[{{node metrics/auc/assert_greater_equal/Assert/AssertGuard/else/_24/Assert}}]]
0 successful operations.
0 derived errors ignored. [Op:__inference_distributed_function_1619]

Function call stack:
distributed_function -> distributed_function


## Observation(s):
### 1) The above case of using SELU activation and a Non-Adaptive optimizer Nestorov, resulted in error during the very first trial, substantiating the foregone conclusion that SELU activation is indeed , not the right activation function for this dataset.